File: appRvSample/src/main/java/in/til/yogeshkumar/timesanimation/Utils/ScaleInBottomAnimator.java
package in.til.yogeshkumar.timesanimation.Utils;

import android.animation.Animator;
import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;
import android.view.View;
import android.view.animation.AccelerateDecelerateInterpolator;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Yogesh Kumar.
 */
public class ScaleInBottomAnimator extends RecyclerView.ItemAnimator {

    List<RecyclerView.ViewHolder> mPendingAdd = new ArrayList<RecyclerView.ViewHolder>();
    List<RecyclerView.ViewHolder> mPendingRemove = new ArrayList<RecyclerView.ViewHolder>();

    @Override
    public boolean animateDisappearance(@NonNull RecyclerView.ViewHolder viewHolder, @NonNull ItemHolderInfo preLayoutInfo, @Nullable ItemHolderInfo postLayoutInfo) {
        mPendingRemove.add(viewHolder);
        return false;
    }

    @Override
    public boolean animateAppearance(@NonNull RecyclerView.ViewHolder viewHolder, @Nullable ItemHolderInfo preLayoutInfo, @NonNull ItemHolderInfo postLayoutInfo) {
        viewHolder.itemView.setAlpha(0.0f);
        return mPendingAdd.add(viewHolder);
    }

    @Override
    public boolean animatePersistence(@NonNull RecyclerView.ViewHolder viewHolder, @NonNull ItemHolderInfo preLayoutInfo, @NonNull ItemHolderInfo postLayoutInfo) {
        return false;
    }

    @Override
    public boolean animateChange(@NonNull RecyclerView.ViewHolder oldHolder, @NonNull RecyclerView.ViewHolder newHolder, @NonNull ItemHolderInfo preLayoutInfo, @NonNull ItemHolderInfo postLayoutInfo) {
        return false;
    }

    @Override
    public void runPendingAnimations() {
        int animationDuration = 300;
        if (!mPendingAdd.isEmpty()) {
            for (final RecyclerView.ViewHolder viewHolder : mPendingAdd) {
                View target = viewHolder.itemView;
                target.setPivotX(target.getMeasuredWidth() / 2);
                target.setPivotY(target.getMeasuredHeight() / 2);

                AnimatorSet animator = new AnimatorSet();

                animator.playTogether(
                        ObjectAnimator.ofFloat(target, "translationX", -target.getMeasuredWidth(), 0.0f),
                        ObjectAnimator.ofFloat(target, "alpha", target.getAlpha(), 1.0f)
                );

                animator.setTarget(target);
                animator.setDuration(animationDuration);
                animator.setInterpolator(new AccelerateDecelerateInterpolator());
                animator.setStartDelay((animationDuration * viewHolder.getPosition()) / 10);
                animator.addListener(new Animator.AnimatorListener() {
                    @Override
                    public void onAnimationStart(Animator animation) {

                    }

                    @Override
                    public void onAnimationEnd(Animator animation) {
                        mPendingAdd.remove(viewHolder);
                    }

                    @Override
                    public void onAnimationCancel(Animator animation) {

                    }

                    @Override
                    public void onAnimationRepeat(Animator animation) {

                    }
                });
                animator.start();
            }
        }
        if (!mPendingRemove.isEmpty()) {
            for (final RecyclerView.ViewHolder viewHolder : mPendingRemove) {
                View target = viewHolder.itemView;
                target.setPivotX(target.getMeasuredWidth() / 2);
                target.setPivotY(target.getMeasuredHeight() / 2);

                AnimatorSet animator = new AnimatorSet();

                animator.playTogether(
                        ObjectAnimator.ofFloat(target, "translationX", 0.0f, target.getMeasuredWidth()),
                        ObjectAnimator.ofFloat(target, "alpha", target.getAlpha(), 0.0f)
                );

                animator.setTarget(target);
                animator.setDuration(animationDuration);
                animator.setInterpolator(new AccelerateDecelerateInterpolator());
                animator.setStartDelay((animationDuration * viewHolder.getPosition()) / 10);
                animator.start();
            }
        }
    }

    @Override
    public void endAnimation(RecyclerView.ViewHolder viewHolder) {
    }

    @Override
    public void endAnimations() {
    }

    @Override
    public boolean isRunning() {
        return !mPendingAdd.isEmpty() || !mPendingRemove.isEmpty();
    }
}

File: appRvSample/src/main/java/com/timesanimation/Utils/ScaleInBottomAnimator.kt
package com.timesanimation.Utils

import android.animation.Animator
import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.view.animation.AccelerateDecelerateInterpolator
import androidx.recyclerview.widget.RecyclerView
import java.util.*

/**
 * Created by Yogesh Kumar.
 */
class ScaleInBottomAnimator : RecyclerView.ItemAnimator() {
    var mPendingAdd: MutableList<RecyclerView.ViewHolder> = ArrayList()
    var mPendingRemove: MutableList<RecyclerView.ViewHolder> = ArrayList()
    override fun animateDisappearance(viewHolder: RecyclerView.ViewHolder, preLayoutInfo: ItemHolderInfo, postLayoutInfo: ItemHolderInfo?): Boolean {
        mPendingRemove.add(viewHolder)
        return false
    }

    override fun animateAppearance(viewHolder: RecyclerView.ViewHolder, preLayoutInfo: ItemHolderInfo?, postLayoutInfo: ItemHolderInfo): Boolean {
        viewHolder.itemView.alpha = 0.0f
        return mPendingAdd.add(viewHolder)
    }

    override fun animatePersistence(viewHolder: RecyclerView.ViewHolder, preLayoutInfo: ItemHolderInfo, postLayoutInfo: ItemHolderInfo): Boolean {
        return false
    }

    override fun animateChange(oldHolder: RecyclerView.ViewHolder, newHolder: RecyclerView.ViewHolder, preLayoutInfo: ItemHolderInfo, postLayoutInfo: ItemHolderInfo): Boolean {
        return false
    }

    override fun runPendingAnimations() {
        val animationDuration = 300
        if (!mPendingAdd.isEmpty()) {
            for (viewHolder in mPendingAdd) {
                val target = viewHolder.itemView
                target.pivotX = target.measuredWidth / 2.toFloat()
                target.pivotY = target.measuredHeight / 2.toFloat()
                val animator = AnimatorSet()
                animator.playTogether(
                        ObjectAnimator.ofFloat(target, "translationX", -target.measuredWidth.toFloat(), 0.0f),
                        ObjectAnimator.ofFloat(target, "alpha", target.alpha, 1.0f)
                )
                animator.setTarget(target)
                animator.duration = animationDuration.toLong()
                animator.interpolator = AccelerateDecelerateInterpolator()
                animator.startDelay = animationDuration * viewHolder.position / 10.toLong()
                animator.addListener(object : Animator.AnimatorListener {
                    override fun onAnimationStart(animation: Animator) {}
                    override fun onAnimationEnd(animation: Animator) {
                        mPendingAdd.remove(viewHolder)
                    }

                    override fun onAnimationCancel(animation: Animator) {}
                    override fun onAnimationRepeat(animation: Animator) {}
                })
                animator.start()
            }
        }
        if (!mPendingRemove.isEmpty()) {
            for (viewHolder in mPendingRemove) {
                val target = viewHolder.itemView
                target.pivotX = target.measuredWidth / 2.toFloat()
                target.pivotY = target.measuredHeight / 2.toFloat()
                val animator = AnimatorSet()
                animator.playTogether(
                        ObjectAnimator.ofFloat(target, "translationX", 0.0f, target.measuredWidth.toFloat()),
                        ObjectAnimator.ofFloat(target, "alpha", target.alpha, 0.0f)
                )
                animator.setTarget(target)
                animator.duration = animationDuration.toLong()
                animator.interpolator = AccelerateDecelerateInterpolator()
                animator.startDelay = animationDuration * viewHolder.position / 10.toLong()
                animator.start()
            }
        }
    }

    override fun endAnimation(viewHolder: RecyclerView.ViewHolder) {}
    override fun endAnimations() {}
    override fun isRunning(): Boolean {
        return !mPendingAdd.isEmpty() || !mPendingRemove.isEmpty()
    }
}